### PR TITLE
Map variables for linen

### PR DESCRIPTION
**Pull request by jheek**
- refactor old transform and transform_module to map_variables
- add map_variables to linen
- add tied autoencoder test to linen transforms

### CODE REVIEW

File: flax/core/lift.py [hunk 1]
```diff
@@ -215,84 +216,48 @@ def repack(inner_scope_tree):
...
+    map_in_fn: Callable[..., Any] = id_fn,
+    map_out_fn: Callable[..., Any] = id_fn,
     init: bool = False, mutable: bool = False,
     rngs: PRNGSequenceFilter = True, variables: CollectionFilter = True):
-  """Locally transform Variables inside a scope.
+  """Locally map Variables inside a scope.
 
   Args:
     fn: the function to be transformed.
     target: the collection(s) to be transformed.
```
Comment (levskaya) [inline flax/core/lift.py:230]: target -> mapped_collections?

here and below I see both `target` and `mapped_collections` used but clearly only one makes sense.

File: flax/core/lift.py [hunk 2]
```diff
@@ -215,84 +216,48 @@ def repack(inner_scope_tree):
...
-    trans_in_fn: creates a view of the target variables.
-    trans_out_fn: transforms the updated variables in the view after mutation.
+    map_in_fn: creates a view of the target variables.
+    map_out_fn: transforms the updated variables in the view after mutation.
     init: If True, variables are initialized before transformation.
     rngs: PRNGSequences added to the transformed scope (default: all).
```
Comment (levskaya) [inline flax/core/lift.py:234]: missing a `mutable` arg here?

File: flax/core/lift.py [hunk 3]
```diff
@@ -215,84 +216,48 @@ def repack(inner_scope_tree):
...
+    scopes = scope_fn((target, variables), rng_groups)
+    y = fn(scopes, *args, **kwargs)
+    out_target, out_vars = repack(scopes)
     if is_target_out:
-      out_target_tree = trans_out_fn(treedef.unflatten(out_target))
-      out_target = treedef.flatten_up_to(out_target_tree)
+      out_target = tuple(map_out_fn(x) for x in out_target)
     return y, (out_target, out_vars)
 
   is_target_out = mutable or init
```
Comment (levskaya) [inline flax/core/lift.py:257]: move `is_target_out` definition above its uses in conditionals ^ ^ ^

File: flax/linen/transforms.py [hunk 4]
```diff
@@ -641,6 +641,24 @@ def __call__(self, c, xs):
       methods=methods)
 
 
+def map_variables(
+    target: Target,
+    mapped_collections: lift.CollectionFilter,
+    trans_in_fn: Callable[..., Any] = lift.id_fn,
+    trans_out_fn: Callable[..., Any] = lift.id_fn,
+    init: bool = False, mutable: bool = False,
+    rngs: lift.PRNGSequenceFilter = True,
```
Comment (levskaya) [inline flax/linen/transforms.py:650]: again, missing `mutable`?

**Review by levskaya — APPROVED** [review]
looks good over all - but I'm wondering if you forgot to commit some working changes, there's a bunch of things that look broken in the map_variables fn?

File: flax/core/lift.py (hunk 3 above)
Comment (jheek) [reply flax/core/lift.py:257]: is_target_out is used outside the wrapper as well to determine the out_variable_groups

Comment (levskaya) [reply flax/core/lift.py:257]: oh sure, I know it works in scope, I was just suggesting lexically to move it to the top before any uses of it.

Comment (jheek) [reply flax/core/lift.py:258]: Fair enough

**Pull request merged**